Adaptions of JobManager after Discussion (#564)
kyma-project/community: internal/proposals/migration-hooks/migration-hooks-proposal.md

- It should be easy to mark a job at which certain point it should be deprecated. Written "by hand" or using some techonology to let pipelines fail, if some jobs exist which should be deprecated.
- JobManager only supports Kyma deploy and not uninstall, to avoid the situation in which developer misuse jobs to clean up dirty left-overs from `kyma uninstall`.

- This mechanism supports jobs for two different use cases: The __component-based__ jobs and the __global/component-independent__ jobs
  - __Component-based__:
    - Check whether the component is installed on the cluster or must be newly installed; and only trigger if it must be installed.
    - It should be possible to trigger jobs before and after a deployment of a component.
  - __Global / Component-independent__:
    - Always trigger the job when installing or upgrading Kyma.
    - It should be possible to trigger jobs before and after the deployment of Kyma.
    - Call component-independent jobs `global` jobs to stick to the naming convention of our helm charts.

### Possible Solution

> To fulfill the requirements, a new package, called `JobManager`, is introduced, which registers, manages, and triggers certain jobs to have a fully-automated installation, or migration. This package has two (hash)maps to manage the workload: One for `pre`-jobs and one for `post`-jobs. In the (hash)maps, the key is the name of the component the jobs belong to, and the value is a slice of the jobs.

Reviewer: ```suggestion
To fulfill the requirements, a new package, called `JobManager`, is introduced, which registers, manages, and triggers certain jobs to have a fully-automated installation or migration. This package has two (hash)maps to manage the workload: One for `pre`-jobs and one for `post`-jobs. In the (hash)maps, the key is the name of the component the jobs belong to, and the value is a slice of the jobs.
```

(for a list of only 2 items, no oxford comma is needed :))

Furthermore, the `JobManager` package has a `duration` variable for benchmarking.

Jobs are implemented within the `JobManager` package in `go`-files, one for each component, using the specific `job`-interface. Then, the implemented interface is registered using `register(job)` in the same file.

The JobManager is used by the `deployment` package in the `deployment.go` file. At the hooks, during the deployment phase, each hook only has to check if the key for the wanted component is present in the pre/post-map. If it's present, the jobs in the map are trigged, if not, nothing must be done.

To benchmark the jobs, a timer is used in the pre- and post-job triggers.

Retries for the jobs are not handled by the JobManager. Retries should be implemented by the jobs themselves, because it's more flexible and the interface is easy to manage. Also, the check if the logic of the job should be executed stays inside of the job, and is not implemented by the JobManager.

### Deprecation of Jobs

Passive:

> - Tag at which certain Kyma version jobs should be deprecated as a comment above the register call.

Reviewer: Are the jobs deprecated as a comment? Or is the tag a comment? Assuming the second meaning.

```suggestion
- Tag (as a comment above the register call) at which defined Kyma version the jobs should be deprecated.
```

Active:

> - Go-Build-Tags cannot be used for this usecase.

Reviewer: ```suggestion
- Go-Build-Tags cannot be used for this use case.
```

> - Add `deprecation` function to job-interface, which returns at which Kyma-version job should be deprecated. Before job is executed, deprecation fucntion is called to check if it is already deprecaded, if yes an Error should be thrown to block the CI.

Reviewer: ```suggestion
- Add `deprecation` function to job-interface, which returns at which Kyma version the job should be deprecated. Before the job is executed, the deprecation function is called to check whether it is already deprecated. If deprecated, an Error should be thrown to block the CI.
```

Furthermore, the `JobManager` package has a `duration` variable for benchmarking.

> Jobs are implemented within the `JobManager` package in `go`-files, one for each component, using the specific `job`-interface. Then, the implemented interface is registered using `register(job)` in the same file. To implement the `job`-interface, the newly created jobs need to implement the `execute(*config.Config, kubernetes.Interface)` function, which takes the installation config and an kubernetes interface as input, so that the jobs can interact with the cluster. The return value needs to be an error. Additionally the `when()` function needs to be implemented, which returns the component the job is bound to and if it should run pre or post the deployment. And if the active solution for tagging jobs as deprecated is choosen, then the `deprecate` function also needs to be implemented - more in the next section.

Reviewer: ```suggestion
Jobs are implemented within the `JobManager` package in `go`-files, one for each component, using the specific `job` interface. Then, the implemented interface is registered using `register(job)` in the same file.
To implement the `job` interface, the newly created jobs must implement the `execute(*config.Config, kubernetes.Interface)` function, which takes the installation config and a kubernetes interface as input, so that the jobs can interact with the cluster. The return value must be an error. Additionally, the `when()` function must be implemented, which returns the component the job is bound to and whether it should run pre or post the deployment. If the active solution for tagging jobs as deprecated is chosen, then the `deprecate` function also must be implemented - more in the next section.
```

register(job1)
type job1 struct{}

> func (j job1) execute(cfg *config.Config, kubeClient kubernetes.Interface) {

Reviewer 2: I think we also seed cancellation context here. But maybe it's too technical/low-level to introduce in the proposal?
